### ISSUE TITLE
a question about quantity imbalance

### ISSUE THREAD

**Issue by hejun888** [elcorto/imagecluster#18]
The 4000 images were clustered, but there were more than 3000 images in one category and less in other categories, resulting in quantity imbalance.   I want to ask how do you make the number of categories not too different.

**Comment by elcorto** [comment]
The number of clusters and thus images per cluster depends on the [`sim` parameter](https://elcorto.github.io/imagecluster/generated/api/imagecluster.calc.cluster.html#imagecluster.calc.cluster) as well as on the quality of feature vectors extracted from the embedding network, [currently VGG16 which is hard-coded](https://github.com/elcorto/imagecluster/blob/master/imagecluster/calc.py#L16).

In case you *know* that images are not supposed to be in the same cluster, try changing `sim` and see if they separate.

It may also be that the VGG16 features do not capture the content of your images in a way which aligns with your prior knowledge, resulting in images being too close in feature space. After all, using a pre-trained embedding network is just a simple hack that often works well enough in practice where the assumption is that new images are drawn from a distribution not terribly different from the one used to train it. You may try a different embedding net. You can also train an auto-encoder on your dataset and cluster using its latent space.

**Comment by hejun888** [comment]
thanks for your reply. I set sim to 0.3, 0.4, 0.5, and the result is similar. How about k-mean.

**Comment by elcorto** [comment]
Sure, k-means is an option, or any other clustering method for that matter, apart from changing how feature vectors ("fingerprints") are computed, as I mentioned above.

You can grab all fingerprints as an array of shape `(n_samples, n_features)` and feed that to e.g. sklearn's k-means. Have a look at `examples/example_api_detailed.py`, replace the line `clusters = calc.cluster(...)` with something like

```py
X = np.array(list(fingerprints.values()))
labels = KMeans(n_clusters=...).fit_predict(X)
```

To use the lines after that using `clusters`, you need to map `labels` to file names somehow. Check how `imagecluster.calc.cluster` does it and adapt to the output format of `KMeans`. I haven't done this so far, so I'd have to look into that myself.

Also you will need to ignore time distance scaling since that is implemented as acting on the feature space distances directly in `imagecluster.calc.cluster`. If you want that, you need to find a way to feed distances instead of feature vectors to the clustering method of your choice. I haven't checked if sklearn supports that.

Hope that helps.

**Comment by hejun888** [comment]
ok，thanks
